nncf/composite_compression.py

```diff
@@ -78,6 +82,71 @@ def load_state_dict(self, state_dict):
             child_scheduler.load_state_dict(state_dict)
 
 
+class CompositeCompressionAlgorithmBuilder(CompressionAlgorithmBuilder):
+    def __init__(self, config: 'NNCFConfig', should_init: bool = True):
+        from nncf import NNCFConfig
+        from nncf.quantization.structs import QuantizerSetupType
+        from nncf.model_creation import get_compression_algorithm
+
+        super().__init__(config, should_init)
+        self._child_builders = []  # type: List[CompressionAlgorithmBuilder]
+
+        compression_config_json_section = config.get('compression', {})
+        compression_config_json_section = deepcopy(compression_config_json_section)
+
+        hw_config_type = None
+        quantizer_setup_type_str = config.get("quantizer_setup_type", "propagation_based")
+        quantizer_setup_type = QuantizerSetupType.from_str(quantizer_setup_type_str)
+        if quantizer_setup_type == QuantizerSetupType.PROPAGATION_BASED:
+            target_device = config.get("target_device", "ANY")
+            if target_device != 'TRIAL':
+                hw_config_type = HWConfigType.from_str(HW_CONFIG_TYPE_TARGET_DEVICE_MAP[target_device])
+
+        if isinstance(compression_config_json_section, dict):
+            compression_config = NNCFConfig(compression_config_json_section)
+            compression_config.register_extra_structs(config.get_all_extra_structs_for_copy())
+            compression_config["hw_config_type"] = hw_config_type
+            compression_config['quantizer_setup_type'] = quantizer_setup_type
+            self._child_builders = [
+                get_compression_algorithm(compression_config)(compression_config, should_init=should_init), ]
+        else:
+            for algo_config in compression_config_json_section:
+                algo_config = NNCFConfig(algo_config)
+                algo_config.register_extra_structs(config.get_all_extra_structs_for_copy())
+                algo_config["hw_config_type"] = hw_config_type
+                algo_config['quantizer_setup_type'] = quantizer_setup_type
+                self._child_builders.append(
+                    get_compression_algorithm(algo_config)(algo_config, should_init=should_init))
+
+    def __bool__(self):
+        return bool(self.child_builders)
+
+    @property
+    def child_builders(self):
+        return self._child_builders
+
+    def apply_to(self, target_model: NNCFNetwork) -> NNCFNetwork:
+        for ctrl in self._child_builders:
+            target_model = ctrl.apply_to(target_model)
+        return target_model
+
+    def handle_frozen_layers(self):
```

Comment:
With this kind of a solution, a single-builder NNCF usage such as:
```
builder = MyBuilder()
builder.apply_to(nncf_network)
ctrl = nncf_network.commit_compression_changes()
```
 won't be handling frozen layers at all. 

I think it would be safe to instead call `_handle_frozen_layers()` for each of the child_builders. When you move the call itself to `apply_to`, implementing a separate `_handle_frozen_layers()` for the composite builder will not be necessary.

Reply:
done